### PR TITLE
Refresh token automatically

### LINKED ISSUE #7: Better support for refresh tokens
### Describe the problem you'd like to have solved

We now have the capability to store a `refresh_token` in the session using `storeRefreshToken`. However, as a developer I currently need to do a lot of heavy lifting to put this to good use:

 - I use the `access_token`
 - When the API I'm calling returns a `401`, I can assume that the token is expired and use the refresh token instead.
 - I need to perform the refresh token exchange myself
 - I don't really have a capability to update the access_token in the session

### Describe the ideal solution

Ideally I would be able to retrieve an access token from the session and everything is handled for me. If the token is valid, return that. If the token is missing a scope, return null. If the token is for an other audience, return null. If the token is expired, renew it, ...

```
const { user } = await auth0.getSession(req, res);
const accessToken = auth0.getAccessToken({
  scope: 'read:calendar',
  audience: 'https://myapp.com/api'
});

### PR DESCRIPTION
**Pull request by karaggeorge**
By submitting a PR to this repository, you agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md). Please see the [contributing guidelines](https://github.com/auth0/open-source-template/blob/master/CONTRIBUTING.md) for how to create and submit a high-quality PR for this repo.

### Description

Fixes #7 (part of it)

When the refresh_token is saved in the session, at any point we try to read the session from the cookie, if the token has expired (also saving expires_at), we refresh the token, save the new session and return it.

The rest of the code doesn't need to have any idea of what happened

#### Parts that weren't implemented:

>  If the token is missing a scope, return null. If the token is for an other audience, return null.

I only focused on the refresh token functionality, no scope/audience checks

### Testing

- [x] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [ ] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [x] All active GitHub checks for tests, formatting, and security are passing
- [ ] The correct base branch is being used, if not `master`

----

I haven't added doc changes yet. I'm not sure if this needs its own setting (`autoRefresh`?) or if it's just something built in.

There wasn't much discussion on the issue, so I'll make the doc updates when I hear more here

### CODE REVIEW

**Comment by nodabladam** [comment]
@sandrinodimattia Could you please merge this and 38 into a release?  If it can at least get into an alpha build we would like to help test it.  Nextjs-auth0 seems great and my company would like to use it but when testing it, these are total blockers for us.

**Comment by nodabladam** [comment]
@karaggeorge When doing the build for this, I am getting 
```

src/instance.node.ts:26:5 - error TS2322: Type '(req: NextApiRequest, res: NextApiResponse<any>) => Promise<ISession | null | undefined>' is not assignable to type '(req: IncomingMessage) => Promise<ISession | null | undefined>'.

26     getSession: handlers.SessionHandler(store),
       ~~~~~~~~~~

  src/instance.ts:32:3
    32   getSession: (req: IncomingMessage) => Promise<ISession | null | undefined>;
         ~~~~~~~~~~
    The expected type comes from property 'getSession' which is declared here on type 'ISignInWithAuth0'


```

**Comment by karaggeorge** [comment]
@nodabladam Ah, missed that. I'll push an update tomorrow morning when I get to my computer

**Comment by nodabladam** [comment]
@karaggeorge  When running the tests I get an error about `save(): Promise<void>` in profile.test.js being incorrect

should store.ts 18 become `Promise<ISession | null | undefined>` like the read above it:
```
save(req: IncomingMessage, res: ServerResponse, session: ISession): Promise<ISession | null | undefined>;
```
and profile.test.ts 14
```
save(): Promise<ISession | null | undefined> {
   return Promise.resolve(session);
}
```
?

**Comment by karaggeorge** [comment]
@sandrinodimattia @nodabladam Fixed the typing issue and the tests 👍

**Comment by damiannodabl** [comment]
@karaggeorge is this storing the refresh_token in the browser's cookie?, and sure is encrypted by cookieSecret, but wonder how secure could be

**Comment by nodabladam** [comment]
I am totally new to auth0 but since next.js is SPA other than initial renders, would it not make sense to default to silent auth? https://auth0.com/docs/api-auth/tutorials/silent-authentication

**Comment by karaggeorge** [comment]
@damiannodabl the refresh token was stored in the browser's cookie when this library came out. I didn't change that, I just used it. If that's something that needs to change, it'll probably need a new PR.

@nodabladam I haven't used that, but I can look into it. To be honest, I don't currently use Auth0 (or have ever used it), but my current project is a Next.js app that uses Azure AD for authentication, which follows a very similar open id protocol, and I had a hard time figuring out how to set it up, until I read the article about this package. My implementation now, is very similar and very heavily inspired by this code, so when I added these few things to my project, I figured I could open some PRs and try adding them here.

TLDR: If that's something people want, I can try reading the docs on that and maybe adding an alternative way of refreshing. Sounds like it would be more complex, since you'd have to redirect to login -> auth0 -> callback and then make your initial call again, versus this which refreshes it in the background

**Comment by nodabladam** [comment]
@karaggeorge Thanks for your feedback here.  I think what you are saying makes sense and thank you so much for your contributions!

**Comment by tsaiDavid** [comment]
@karaggeorge - this looks great!

@sandrinodimattia - what else is required to help move this forward? My team really needs to be able to easily handle refresh tokens. :)

**Comment by faisalil** [comment]
@sandrinodimattia and @nodabladam , from what I am reading this change is looking good and its only waiting for the merge. Any chance we can do that and kick of a release? there is other open PRs that I am not sure has more clear timelines.

**Comment by nodabladam** [comment]
Our app can't function without 37 and 38 so we have been using them on a hacked fork for a few weeks.  So far so good.  It would be great to get these merged in even if it was an alpha version.

**Comment by sandrinodimattia** [comment]
Thanks for all of this work @karaggeorge. I've used this as the foundation for the work in release v0.8.0. The implementation is however different. In this PR we are mixing concerns of session management (the cookie store) and token renewal, while the v0.8.0 release introduces a new concept, the `TokenCache` which can give you a valid access token (and refresh if necessary + persist in the session).

The change can be found here: https://github.com/auth0/nextjs-auth0/commit/4dd5cdabba5b677a1c078a20e4c1a55ae27e21c8

And a full example is here: https://github.com/auth0/nextjs-auth0/tree/master/examples/api-call-example

I'm going to go ahead and close this PR, thanks again for your work @karaggeorge!
